File: bindings/go/engine_test.go
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://github.com/openpitkit and the OWNERS file for details.

package openpit

import (
	"errors"
	"strings"
	"testing"

	"go.openpit.dev/openpit/accountadjustment"
	"go.openpit.dev/openpit/internal/native"
	"go.openpit.dev/openpit/model"
	"go.openpit.dev/openpit/param"
	"go.openpit.dev/openpit/pretrade"
	"go.openpit.dev/openpit/reject"
	"go.openpit.dev/openpit/tx"
)

func TestEngineBuilderCloseIsIdempotent(*testing.T) {
	builder := NewEngineBuilder().
		WithFullSync().
		CheckPreTradeStartPolicy(&engineTestStartPolicy{name: "p"})

	builder.Close()
	builder.Close()
}

func TestEngineBuilderBuildFailsAfterClose(t *testing.T) {
	builder := NewEngineBuilder().
		WithFullSync().
		CheckPreTradeStartPolicy(&engineTestStartPolicy{name: "p"})

	builder.Close()
	engine, err := builder.Build()
	if engine != nil {
		engine.Stop()
		t.Fatal("Build() engine != nil, want nil")
	}
	if err == nil {
		t.Fatal("Build() error = nil, want non-nil")
	}
}

func TestEngineBuilderScheduleCloseAfterPolicyAddFailure(t *testing.T) {
	second := &engineTestStartPolicy{name: "second"}

	// A forced-failure applier triggers an error on the first Builtin call.
	// Subsequent policy adds must see the error and schedule policies for
	// cleanup.
	builder := NewEngineBuilder().WithFullSync().
		Builtin(&engineTestFailingBuilder{})
	builder.CheckPreTradeStartPolicy(second)

	if second.closeCalls != 0 {
		t.Fatalf("second closeCalls before Build() = %d, want 0", second.closeCalls)
	}

	_, err := builder.Build()
	if err == nil {
		t.Fatal("Build() error = nil, want non-nil")
	}
	if !strings.Contains(err.Error(), "forced") {
		t.Fatalf(
			"Build() error = %q, want to contain %q",
			err.Error(), "forced",
		)
	}
	if second.closeCalls != 1 {
		t.Fatalf("second closeCalls after Build() = %d, want 1", second.closeCalls)
	}
}

func TestEngineBuilderPolicyAddErrorFormatsMessage(t *testing.T) {
	err := newEngineBuilderPolicyAddError(errors.New("forced"), "policy-a")
	if got, want := err.Error(), `failed to add policy "policy-a": forced`; got != want {
		t.Fatalf("Error() = %q, want %q", got, want)
	}
}

func TestEngineStartPreTradeReturnsErrorAfterStop(t *testing.T) {
	engine := newEngineForTests(t)
	engine.Stop()

	request, rejects, err := engine.StartPreTrade(model.NewOrder())
	if request != nil {
		request.Close()
		t.Fatal("StartPreTrade() request != nil, want nil")
	}
	if rejects != nil {
		t.Fatalf("StartPreTrade() rejects = %v, want nil", rejects)
	}
	if err == nil {
		t.Fatal("StartPreTrade() error = nil, want non-nil")
	}
}

func TestEngineExecutePreTradeReturnsErrorAfterStop(t *testing.T) {
	engine := newEngineForTests(t)
	engine.Stop()

	reservation, rejects, err := engine.ExecutePreTrade(model.NewOrder())
	if reservation != nil {
		reservation.Close()
		t.Fatal("ExecutePreTrade() reservation != nil, want nil")
	}
	if rejects != nil {
		t.Fatalf("ExecutePreTrade() rejects = %v, want nil", rejects)
	}
	if err == nil {
		t.Fatal("ExecutePreTrade() error = nil, want non-nil")
	}
}

func TestEngineApplyAccountAdjustmentEmptyBatchIsNoop(t *testing.T) {
	engine := newEngineForTests(t)
	defer engine.Stop()

	rejects, err := engine.ApplyAccountAdjustment(param.NewAccountIDFromInt(1), nil)
	if err != nil {
		t.Fatalf("ApplyAccountAdjustment() error = %v, want nil", err)
	}
	if rejects.IsSet() {
		t.Fatalf("ApplyAccountAdjustment() rejects = %v, want none", rejects)
	}
}

func TestEngineApplyAccountAdjustmentReturnsBatchReject(t *testing.T) {
	engine, err := NewEngineBuilder().
		WithFullSync().
		AccountAdjustmentPolicy(&engineTestRejectingAdjustmentPolicy{name: "adjustment-reject"}).
		Build()
	if err != nil {
		t.Fatalf("Build() error = %v", err)
	}
	defer engine.Stop()

	rejects, err := engine.ApplyAccountAdjustment(
		param.NewAccountIDFromInt(1),
		[]model.AccountAdjustment{model.NewAccountAdjustment()},
	)
	if err != nil {
		t.Fatalf("ApplyAccountAdjustment() error = %v", err)
	}
	if !rejects.IsSet() {
		t.Fatal("ApplyAccountAdjustment() rejects.IsSet() = false, want true")
	}
	batchReject, ok := rejects.Get()
	if !ok {
		t.Fatal("ApplyAccountAdjustment() rejects.Get() ok = false, want true")
	}
	if batchReject.FailedAdjustmentIndex != 0 {
		t.Fatalf("FailedAdjustmentIndex = %d, want 0", batchReject.FailedAdjustmentIndex)
	}
	if len(batchReject.Rejects) != 1 {
		t.Fatalf("batch reject len = %d, want 1", len(batchReject.Rejects))
	}
	if batchReject.Rejects[0].Policy != "adjustment-reject" {
		t.Fatalf("reject policy = %q, want %q", batchReject.Rejects[0].Policy, "adjustment-reject")
	}
}

func newEngineForTests(t *testing.T) *Engine {
	t.Helper()

	engine, err := NewEngineBuilder().
		WithFullSync().
		CheckPreTradeStartPolicy(&engineTestNoopStartPolicy{}).
		Build()
	if err != nil {
		t.Fatalf("Build() error = %v", err)
	}
	return engine
}

type engineTestStartPolicy struct {
	name       string
	closeCalls int
}

func (p *engineTestStartPolicy) Close() {
	p.closeCalls++
}

func (p engineTestStartPolicy) Name() string {
	return p.name
}

func (engineTestStartPolicy) CheckPreTradeStart(pretrade.Context, model.Order) []reject.Reject {
	return nil
}

func (engineTestStartPolicy) ApplyExecutionReport(model.ExecutionReport) bool {
	return false
}

type engineTestNoopStartPolicy struct{}

func (engineTestNoopStartPolicy) Close() {}

func (engineTestNoopStartPolicy) Name() string { return "noop" }

func (engineTestNoopStartPolicy) CheckPreTradeStart(pretrade.Context, model.Order) []reject.Reject {
	return nil
}

func (engineTestNoopStartPolicy) ApplyExecutionReport(model.ExecutionReport) bool {
	return false
}

type engineTestRejectingAdjustmentPolicy struct {
	name string
}

func (engineTestRejectingAdjustmentPolicy) Close() {}

func (p engineTestRejectingAdjustmentPolicy) Name() string {
	return p.name
}

func (p *engineTestRejectingAdjustmentPolicy) ApplyAccountAdjustment(
	accountadjustment.Context,
	param.AccountID,
	model.AccountAdjustment,
	tx.Mutations,
) []reject.Reject {
	return reject.NewSingleItemList(
		reject.CodeOther,
		p.name,
		"adjustment rejected",
		"rejected in test policy",
		reject.ScopeAccount,
	)
}

type engineTestFailingBuilder struct{}

func (*engineTestFailingBuilder) Build(_ native.EngineBuilder) error {
	return errors.New("forced")
}

File: bindings/cpp/test/engine_test.cpp
// Copyright The Pit Project Owners. All rights reserved.
// SPDX-License-Identifier: Apache-2.0
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please see https://openpit.dev and the OWNERS file for details.

#include "openpit/engine.hpp"

#include "openpit/account_adjustment.hpp"
#include "openpit/model.hpp"
#include "openpit/pretrade/pretrade.hpp"
#include "openpit/reject.hpp"

#include <gtest/gtest.h>

#include <cstdint>
#include <optional>
#include <stdexcept>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

namespace {

using openpit::Engine;
using openpit::EngineBuilder;
using openpit::SyncPolicy;
using openpit::param::AccountId;
using openpit::param::Price;
using openpit::param::Quantity;
using openpit::pretrade::Reject;
using openpit::pretrade::RejectCode;

namespace policies = openpit::pretrade::policies;

// Builds the canonical single-leg test order for `accountId`: a buy of one AAPL
// settled in USD at price 100, mirroring the Go `rateLimitTestOrder` helper.
[[nodiscard]] openpit::model::Order TestOrder(std::uint64_t accountId) {
  openpit::model::Order order;
  openpit::model::OrderOperation op;
  op.instrument = openpit::model::Instrument("AAPL", "USD");
  op.accountId = ::openpit::param::AccountId::FromUint64(accountId);
  op.side = openpit::model::Side::Buy;
  op.tradeAmount =
      openpit::model::TradeAmount::OfQuantity(Quantity::FromString("1"));
  op.price = Price::FromString("100");
  order.operation = std::move(op);
  return order;
}

// A rate-limit engine that admits a single order on the broker axis: the second
// pre-trade for any account is rejected with RateLimitExceeded.
[[nodiscard]] Engine SingleOrderEngine() {
  EngineBuilder builder(SyncPolicy::None);
  policies::RateLimitPolicy config;
  config.BrokerBarrier(policies::RateLimitBrokerBarrier(policies::RateLimit(
      /*maxOrders=*/1, /*windowNanoseconds=*/60'000'000'000)));
  config.AddTo(builder);
  return builder.Build();
}

template <typename Handler>
[[nodiscard]] Engine CustomPolicyEngine(Handler handler) {
  EngineBuilder builder(SyncPolicy::Full);
  openpit::pretrade::CustomPolicy<Handler> policy("ThrowingPolicy",
                                                  std::move(handler));
  builder.Add(policy);
  return builder.Build();
}

template <typename Callable>
[[nodiscard]] std::string ErrorMessageFrom(Callable&& call) {
  try {
    call();
  } catch (const openpit::Error& error) {
    return error.Message();
  }
  return {};
}

class ThrowingStartPolicy {
 public:
  [[nodiscard]] std::string_view Name() const noexcept {
    return "ThrowingStartPolicy";
  }

  [[nodiscard]] std::optional<Reject> CheckPreTradeStart(
      const openpit::Order& /*order*/) const {
    throw std::runtime_error("start callback failed");
  }
};

class ThrowingMainPolicy {
 public:
  [[nodiscard]] std::string_view Name() const noexcept {
    return "ThrowingMainPolicy";
  }

  void PerformPreTradeCheck(
      const openpit::pretrade::Context& /*context*/,
      openpit::pretrade::PolicyDecision& /*decision*/) const {
    throw std::runtime_error("main callback failed");
  }
};

class ThrowingDryRunPolicy {
 public:
  [[nodiscard]] std::string_view Name() const noexcept {
    return "ThrowingDryRunPolicy";
  }

  [[nodiscard]] std::optional<Reject> CheckPreTradeStartDryRun(
      const openpit::Order& /*order*/) const {
    throw 17;
  }
};

class ThrowingReportPolicy {
 public:
  [[nodiscard]] std::string_view Name() const noexcept {
    return "ThrowingReportPolicy";
  }

  [[nodiscard]] bool ApplyExecutionReport(
      const openpit::ExecutionReport& /*report*/) const {
    throw std::runtime_error("report callback failed");
  }
};

//------------------------------------------------------------------------------
// Builder -> build with policies.

TEST(EngineBuilder, BuildsWithBuiltinPolicy) {
  EngineBuilder builder(SyncPolicy::Full);
  builder.Add(policies::OrderValidationPolicy{});
  Engine engine = builder.Build();
  EXPECT_TRUE(static_cast<bool>(engine));
}

TEST(EngineBuilder, BuildWithoutPoliciesThrows) {
  EngineBuilder builder(SyncPolicy::Full);
  EXPECT_THROW({ Engine engine = builder.Build(); }, openpit::Error);
}

TEST(EngineBuilder, DuplicatePolicyGroupIdThrowsBuildError) {
  EngineBuilder builder(SyncPolicy::Full);
  // Two builtin policies forced onto the same non-default group id collide.
  builder.Add(policies::OrderValidationPolicy{}.PolicyGroupId(7));
  builder.Add(policies::RateLimitPolicy{}.PolicyGroupId(7).BrokerBarrier(
      policies::RateLimitBrokerBarrier(
          policies::RateLimit(1, 60'000'000'000))));
  EXPECT_THROW({ Engine engine = builder.Build(); }, openpit::EngineBuildError);
}

//------------------------------------------------------------------------------
// StartPreTrade.

TEST(EngineStartPreTrade, HappyPathReturnsRequest) {
  Engine engine = SingleOrderEngine();

  openpit::pretrade::StartResult result = engine.StartPreTrade(TestOrder(1));
  EXPECT_TRUE(result.Passed());
  EXPECT_TRUE(result.rejects.empty());
  ASSERT_TRUE(result.request.has_value());
  EXPECT_TRUE(static_cast<bool>(*result.request));
}

TEST(EngineStartPreTrade, RejectPathReturnsRejectsNotRequest) {
  Engine engine = SingleOrderEngine();

  // First order consumes the single-order budget.
  openpit::pretrade::StartResult first = engine.StartPreTrade(TestOrder(1));
  ASSERT_TRUE(first.Passed());

  // Second order is rejected by the rate limiter.
  openpit::pretrade::StartResult second = engine.StartPreTrade(TestOrder(1));
  EXPECT_FALSE(second.Passed());
  EXPECT_FALSE(second.request.has_value());
  ASSERT_EQ(second.rejects.size(), 1u);
  EXPECT_EQ(second.rejects.front().code, RejectCode::RateLimitExceeded);
}

TEST(EngineStartPreTrade, AbiFailureThrows) {
  // A default-constructed (null-handle) engine is an invalid pointer to the C
  // ABI: the boundary failure must surface as a thrown openpit::Error, never a
  // reject value.
  const Engine engine;
  EXPECT_THROW(
      { auto result = engine.StartPreTrade(TestOrder(1)); }, openpit::Error);
}

TEST(EngineStartPreTrade, CallbackExceptionThrowsError) {
  Engine engine = CustomPolicyEngine(ThrowingStartPolicy{});

  const std::string message = ErrorMessageFrom(
      [&] { static_cast<void>(engine.StartPreTrade(TestOrder(1))); });

  EXPECT_NE(message.find("start callback failed"), std::string::npos);
}

//------------------------------------------------------------------------------
// Non-mutating dry-run.

TEST(EngineDryRun, StartDryRunDoesNotConsumeRateLimitBudget) {
  Engine engine = SingleOrderEngine();

  const openpit::pretrade::DryRunReport probe =
      engine.StartPreTradeDryRun(TestOrder(1));
  EXPECT_TRUE(probe.Passed());
  EXPECT_TRUE(probe.Rejects().empty());
  EXPECT_TRUE(probe.Lock().IsEmpty());
  EXPECT_TRUE(probe.AccountAdjustments().empty());
  EXPECT_TRUE(probe.AccountBlocks().empty());

  EXPECT_TRUE(engine.StartPreTrade(TestOrder(1)).Passed());
  const openpit::pretrade::StartResult second =
      engine.StartPreTrade(TestOrder(1));
  EXPECT_FALSE(second.Passed());
  ASSERT_EQ(second.rejects.size(), 1u);
  EXPECT_EQ(second.rejects.front().code, RejectCode::RateLimitExceeded);
}

TEST(EngineDryRun, ExecuteDryRunDoesNotConsumeRateLimitBudget) {
  Engine engine = SingleOrderEngine();

  const openpit::pretrade::DryRunReport probe =
      engine.ExecutePreTradeDryRun(TestOrder(1));
  EXPECT_TRUE(probe.Passed());
  EXPECT_TRUE(probe.Rejects().empty());

  const openpit::pretrade::ExecuteResult first =
      engine.ExecutePreTrade(TestOrder(1));
  EXPECT_TRUE(first.Passed());

  const openpit::pretrade::ExecuteResult second =
      engine.ExecutePreTrade(TestOrder(1));
  EXPECT_FALSE(second.Passed());
  ASSERT_EQ(second.rejects.size(), 1u);
  EXPECT_EQ(second.rejects.front().code, RejectCode::RateLimitExceeded);
}

TEST(EngineDryRun, UnknownCallbackExceptionThrowsError) {
  Engine engine = CustomPolicyEngine(ThrowingDryRunPolicy{});

  const std::string message = ErrorMessageFrom(
      [&] { static_cast<void>(engine.StartPreTradeDryRun(TestOrder(1))); });

  EXPECT_EQ(message, "unknown callback error");
}

//------------------------------------------------------------------------------
// ExecutePreTrade + reservation resolution.

TEST(EngineExecutePreTrade, HappyPathReturnsReservation) {
  Engine engine = SingleOrderEngine();

  openpit::pretrade::ExecuteResult result =
      engine.ExecutePreTrade(TestOrder(1));
  EXPECT_TRUE(result.Passed());
  EXPECT_TRUE(result.rejects.empty());
  ASSERT_TRUE(result.reservation.has_value());
  // Resolving the reservation must not throw.
  EXPECT_NO_THROW(result.reservation->Commit());
}

TEST(EngineExecutePreTrade, RollbackDoesNotReleaseRateLimitBudget) {
  Engine engine = SingleOrderEngine();

  // The rate-limit budget is consumed by the start stage, not by the main-stage
  // reservation: every start attempt (even rejected ones) permanently counts in
  // the window. Rolling back the reservation only unwinds main-stage mutations,
  // so it cannot return the consumed rate-limit slot. With a single-order
  // limit, the first execute consumes the slot and a subsequent execute is
  // rejected regardless of the rollback.
  {
    openpit::pretrade::ExecuteResult first =
        engine.ExecutePreTrade(TestOrder(1));
    ASSERT_TRUE(first.reservation.has_value());
    first.reservation->Rollback();
  }
  openpit::pretrade::ExecuteResult second =
      engine.ExecutePreTrade(TestOrder(1));
  EXPECT_FALSE(second.Passed());
  ASSERT_EQ(second.rejects.size(), 1u);
  EXPECT_EQ(second.rejects.front().code, RejectCode::RateLimitExceeded);
}

TEST(EngineExecutePreTrade, CommitConsumesReservedBudget) {
  Engine engine = SingleOrderEngine();

  {
    openpit::pretrade::ExecuteResult first =
        engine.ExecutePreTrade(TestOrder(1));
    ASSERT_TRUE(first.reservation.has_value());
    first.reservation->Commit();
  }
  // The committed first order used the single-order budget; the second rejects.
  openpit::pretrade::ExecuteResult second =
      engine.ExecutePreTrade(TestOrder(1));
  EXPECT_FALSE(second.Passed());
  ASSERT_EQ(second.rejects.size(), 1u);
  EXPECT_EQ(second.rejects.front().code, RejectCode::RateLimitExceeded);
}

TEST(EngineExecutePreTrade, AbiFailureThrows) {
  const Engine engine;
  EXPECT_THROW(
      { auto result = engine.ExecutePreTrade(TestOrder(1)); }, openpit::Error);
}

TEST(EngineExecutePreTrade, CallbackExceptionThrowsError) {
  Engine engine = CustomPolicyEngine(ThrowingMainPolicy{});

  const std::string message = ErrorMessageFrom(
      [&] { static_cast<void>(engine.ExecutePreTrade(TestOrder(1))); });

  EXPECT_NE(message.find("main callback failed"), std::string::npos);
}

//------------------------------------------------------------------------------
// Request::Execute (deferred two-stage flow).

TEST(EngineRequest, ExecutePassesThenCommit) {
  Engine engine = SingleOrderEngine();

  openpit::pretrade::StartResult start = engine.StartPreTrade(TestOrder(1));
  ASSERT_TRUE(start.request.has_value());

  openpit::pretrade::ExecuteResult executed = start.request->Execute();
  EXPECT_TRUE(executed.Passed());
  ASSERT_TRUE(executed.reservation.has_value());
  EXPECT_NO_THROW(executed.reservation->Commit());
}

TEST(EngineRequest, SecondDeferredFlowIsRejectedAtStart) {
  Engine engine = SingleOrderEngine();

  // Drive the first order all the way to commit.
  openpit::pretrade::StartResult first = engine.StartPreTrade(TestOrder(1));
  ASSERT_TRUE(first.request.has_value());
  openpit::pretrade::ExecuteResult firstExec = first.request->Execute();
  ASSERT_TRUE(firstExec.reservation.has_value());
  firstExec.reservation->Commit();

  // The rate limiter runs in the start stage, so the second order is rejected
  // by StartPreTrade itself: no deferred request is produced and the reject is
  // surfaced immediately rather than at the main stage.
  openpit::pretrade::StartResult second = engine.StartPreTrade(TestOrder(1));
  EXPECT_FALSE(second.Passed());
  EXPECT_FALSE(second.request.has_value());
  ASSERT_EQ(second.rejects.size(), 1u);
  EXPECT_EQ(second.rejects.front().code, RejectCode::RateLimitExceeded);
}

TEST(EngineRequest, CallbackExceptionThrowsError) {
  Engine engine = CustomPolicyEngine(ThrowingMainPolicy{});

  openpit::pretrade::StartResult start = engine.StartPreTrade(TestOrder(1));
  ASSERT_TRUE(start.request.has_value());

  const std::string message =
      ErrorMessageFrom([&] { static_cast<void>(start.request->Execute()); });

  EXPECT_NE(message.find("main callback failed"), std::string::npos);
}

//------------------------------------------------------------------------------
// ApplyExecutionReport.

// Builds an execution report carrying the operation identity for `accountId`.
[[nodiscard]] openpit::model::ExecutionReport TestReport(
    std::uint64_t accountId) {
  openpit::model::ExecutionReport report;
  openpit::model::ExecutionReportOperation op;
  op.instrument = openpit::model::Instrument("AAPL", "USD");
  op.accountId = ::openpit::param::AccountId::FromUint64(accountId);
  op.side = openpit::model::Side::Buy;
  report.operation = std::move(op);
  return report;
}

TEST(EngineApplyExecutionReport, OrderValidationProducesNoBlocks) {
  EngineBuilder builder(SyncPolicy::Full);
  builder.Add(policies::OrderValidationPolicy{});
  Engine engine = builder.Build();

  // OrderValidation neither blocks nor produces adjustment outcomes; the report
  // applies cleanly with empty result channels.
  const openpit::PostTradeResult result =
      engine.ApplyExecutionReport(TestReport(1));
  EXPECT_TRUE(result.accountBlocks.empty());
  EXPECT_TRUE(result.accountAdjustmentOutcomes.empty());
}

TEST(EngineApplyExecutionReport, AbiFailureThrows) {
  const Engine engine;
  EXPECT_THROW(
      { auto result = engine.ApplyExecutionReport(TestReport(1)); },
      openpit::Error);
}

TEST(EngineApplyExecutionReport, CallbackExceptionThrowsError) {
  Engine engine = CustomPolicyEngine(ThrowingReportPolicy{});

  const std::string message = ErrorMessageFrom(
      [&] { static_cast<void>(engine.ApplyExecutionReport(TestReport(1))); });

  EXPECT_NE(message.find("report callback failed"), std::string::npos);
}

//------------------------------------------------------------------------------
// ApplyAccountAdjustment.
//
// The adjustment value type is authored by a parallel module
// (openpit/account_adjustment.hpp). The empty-batch path exercises the engine
// method end-to-end without depending on that type: a zero-length batch is
// accepted and applies cleanly. A minimal local stub satisfies the template's
// `Raw()` requirement; it is never invoked for an empty batch.

struct StubAdjustment {
  [[nodiscard]] OpenPitAccountAdjustment Raw() const noexcept {
    return OpenPitAccountAdjustment{};
  }
};

TEST(EngineApplyAccountAdjustment, EmptyBatchApplies) {
  EngineBuilder builder(SyncPolicy::Full);
  builder.Add(policies::OrderValidationPolicy{});
  Engine engine = builder.Build();

  const openpit::AdjustmentResult result =
      engine.ApplyAccountAdjustment<StubAdjustment>(AccountId::FromUint64(1),
                                                    /*adjustments=*/{});
  EXPECT_TRUE(result.Passed());
  EXPECT_TRUE(result.accountAdjustmentOutcomes.empty());
}

TEST(EngineApplyAccountAdjustment, AbiFailureThrows) {
  const Engine engine;
  EXPECT_THROW(
      {
        auto result = engine.ApplyAccountAdjustment<StubAdjustment>(
            AccountId::FromUint64(1), /*adjustments=*/{});
        static_cast<void>(result);
      },
      openpit::Error);
}

//------------------------------------------------------------------------------
// Account currency and runtime configuration.

TEST(EngineAccountCurrency, SetAndClearAccountAndGroupCurrency) {
  EngineBuilder builder(SyncPolicy::Full);
  builder.Add(policies::OrderValidationPolicy{});
  Engine engine = builder.Build();

  const AccountId account = AccountId::FromUint64(42);
  const openpit::param::AccountGroupId group =
      openpit::param::AccountGroupId::FromUint32(7);

  EXPECT_NO_THROW(engine.SetAccountCurrency(account, "USD"));
  EXPECT_NO_THROW(engine.ClearAccountCurrency(account));

  ASSERT_FALSE(engine.Accounts().RegisterGroup({account}, group).has_value());
  EXPECT_NO_THROW(engine.SetAccountGroupCurrency(group, "USD"));
  EXPECT_NO_THROW(engine.ClearAccountGroupCurrency(group));
}

TEST(EngineConfigure, RateLimitUpdateChangesRuntimeBudget) {
  Engine engine = SingleOrderEngine();

  engine.Configure().RateLimit(
      policies::RateLimitPolicyName,
      policies::RateLimitBrokerBarrier(policies::RateLimit(
          /*maxOrders=*/2, /*windowNanoseconds=*/60'000'000'000)));

  EXPECT_TRUE(engine.StartPreTrade(TestOrder(1)).Passed());
  EXPECT_TRUE(engine.StartPreTrade(TestOrder(1)).Passed());

  const openpit::pretrade::StartResult third =
      engine.StartPreTrade(TestOrder(1));
  EXPECT_FALSE(third.Passed());
  ASSERT_EQ(third.rejects.size(), 1u);
  EXPECT_EQ(third.rejects.front().code, RejectCode::RateLimitExceeded);
}

TEST(EngineConfigure, UnknownPolicyThrowsStructuredConfigureError) {
  Engine engine = SingleOrderEngine();

  try {
    engine.Configure().RateLimit(
        "MissingPolicy",
        policies::RateLimitBrokerBarrier(policies::RateLimit(
            /*maxOrders=*/2, /*windowNanoseconds=*/60'000'000'000)));
    FAIL() << "Configure().RateLimit should have thrown";
  } catch (const openpit::ConfigureError& error) {
    EXPECT_EQ(error.Kind(), openpit::ConfigureErrorKind::Unknown);
  }
}

TEST(EngineConfigure, SpotFundsLimitModeUpdateBuildsThroughAccessor) {
  EngineBuilder builder(SyncPolicy::Full);
  builder.Add(policies::SpotFundsPolicy{});
  Engine engine = builder.Build();

  EXPECT_NO_THROW(engine.Configure().SpotFundsGlobalLimitMode(
      policies::SpotFundsPolicyName, policies::SpotFundsLimitMode::TrackOnly));
}

//------------------------------------------------------------------------------
// Account-group lookup via the engine read query.

TEST(EngineAccountGroup, AbsentForUngroupedAccount) {
  EngineBuilder builder(SyncPolicy::Full);
  builder.Add(policies::OrderValidationPolicy{});
  Engine engine = builder.Build();

  EXPECT_FALSE(engine.AccountGroup(AccountId::FromUint64(1)).has_value());
}

}  // namespace
